packages/styled/types/base.d.ts

```diff
@@ -13,19 +17,24 @@ export {
 
 export { ComponentSelector, Interpolation }
 
-/** Same as StyledOptions but shouldForwardProp must be a type guard */
+/**
+ * Same as StyledOptions but shouldForwardProp must be a type guard (https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates).
+ * Practical sense: you can define and reuse atomic `shouldForwardProp` filters that are strictly bound with some `ForwardedProps` type.
+ */
 export interface FilteringStyledOptions<
-  Props,
+  Props = Record<PropertyKey, any>,
```

Comment:
q: why the `Props` have a different default between `FilteringStyledOptions` and `StyledOptions`?

Reply:
> Props = Record<PropertyKey, any>,

This is just an artifact of the first impl. I'll fix.

> So it seems like the only thing to settle on is the variance type for the `shouldForwardProp`.
 
It _seems_ that `strictFunctionTypes` option is not applied to object `method?: () {}` signature. So if we rewrite the code as suggested above we break everything.

```ts
export interface FilteringStyledOptions<
  Props = AnyRecord,
  ForwardedProps extends keyof Props = keyof Props
> {
  label?: string
  shouldForwardProp?: <P extends UnionToIntersection<ForwardedProps>>(propName: P) => propName is P
  target?: string
}

export interface StyledOptions<Props = AnyRecord> {
  label?: string
  shouldForwardProp?: <P extends keyof UnionToIntersection<Props>>(propName: P) => boolean
  target?: string
}

type AnyRecord = Record<PropertyKey, any>

export type UnionToIntersection<U, K = any> = (
  U extends K ? (k: U) => void : never
  ) extends (k: infer I) => void
  ? I
  : never
```

The main problem: I don't know how to infer param type inside generic, so I cannot add a type assertion that impl fn param extends HTML element attributes.
```
type F = <P extends any extends infer P ? P : never>(p: P) => boolean
// P is unknown
const f: F = (prop: 'style') => true
```
I also tried to handle this in another way: to represent `shouldForwardProp` as a union of fns with all possible signatures. I looks syntactically correct and automatable with recursive mapping, but it requires a quantum computer.

```ts
type T = 'a' | 'b' | 'c'
type F = (p: 'a') => true | (p: 'b') => true | (p: 'c') => true 
| (p: 'a' | 'b') => true | (p: 'b' | 'c') => true | (p: 'a' | 'c') => true 
| (p: 'a' | 'b' | 'c') => 
```

For ~300 attributed for every HTML element type we get: `C(n,k) = n! / k!(n-k)!, k = 1, n=300, C =~3E615`.
https://www.numberempire.com/factorialcalculator.php?number=300

PS Sorry for the late reply.

Reply:
> It seems that strictFunctionTypes option is not applied to object method?: () {} signature

If I understand this correctly then with `strictFunctionTypes: false` every function is bivariant but with `strictFunctionTypes: true` **methods** are still bivariant but every other function becomes contravariant.

I think you might have a typo in this example as it should be `method?(): void` (or similar) to fit the rest of the sentence. I've just wanted to clarify that we have a common understanding of this stuff (where I'm not 100% sure if my understanding is fully correct).

> So if we rewrite the code as suggested above we break everything.

Does "as suggested" mean "change to a function type and thus make it contravariant"? Could you give me an example of what this would break? If it breaks something important then I think we should have a type test for this.

>   shouldForwardProp?: <P extends UnionToIntersection<ForwardedProps>>(propName: P) => propName is P

I'm unsure what's happening here. I understand what `UnionToIntersection` does but I don't see why it was applied here. The result of this type is an intersection and `P` definitely should be constrained to a union of strings (or to just a string type). Could you explain what was the goal here?

Reply:
I've pushed several snapshot branches so you can observe my poor efforts:
* [**snap#3**](https://github.com/antongolub/emotion/tree/snap2) All gneric singatures use `Record<PropertyKey, any>` as default type. Filters types exposes as object fields. The onl[ working.
* [**snap#2**](https://github.com/antongolub/emotion/tree/snap2) `shouldForwardProp` type declared as a generic function + `UnionToIntersecition`
* [**snap#1**](https://github.com/antongolub/emotion/tree/snap1) same as 2 but w/o  `UnionToIntersecition`.

`UnionToIntersection` was placed here to ensure that keyof type includes all the keys of all types of `Props` union:
`ClassAttributes<HTMLDivElement> | HTMLAttributes<HTMLDivElement> → keyof ClassAttributes<HTMLDivElement> | keyof HTMLAttributes<HTMLDivElement>`. Just for experiment.

Reply:
It seems that we ran into some limitation (in understanding?) of the current TS. What I want to get:
```ts
// Declare filter signature that operates the keys of set 'a' | 'b' ... 'z'
type F {
  filter: (p: 'a' | 'b' ... 'z') => bool
}

// Introduce a specific filter, that uses a subset p type
const f: F = {
    filter: (p: 'a' | 'c') => bool
}
// 'a' | 'c' extends 'a' | 'c' ... 'z'
```

Reply:
> I've pushed several snapshot branches so you can observe my poor efforts:

Thank you for those - I'll try to review them this week to get a better feeling of what we are dealing with here.

> UnionToIntersection was placed here to ensure that keyof type includes all the keys of all types of Props union:

Do we need to care about `ClassAttributes` though? This interface only contains `key` and `ref` properties and those can't really b received by `shouldForwardProp` as those are special props in React and it won't ever handle them to us.

> It seems that we ran into some limitation (in understanding?) of the current TS. What I want to get:

Ye, this is impossible - mainly because now you could make such a call `f.filter('z')` and it could potentially crash at runtime as the supplied implementation is not capable of handling `'z'`.

So basically, here, assigning to this type is not valid. Such things could be expressed with some function helpers and stuff, like here:
[TS playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAYlC8UDeAoKUBmBLANsCATgFxQAUYJA5AIaVQA+UlARnY5QMZtMAmlAlAgB8UZgHsxOCNQB2KAL4oUoSFACCMkDACuMjsCxiZCMgDpz1AgHMAziVkgA2gF1B8EQ6UroANWo4sHmp8AB4AFSgIAA98GR4bZExcfGJ1TR09AyMoeRFEGEcAImw8QkLnSJiIOISwopKU8qgAfigIkhkIADdCJR4IDhxLaAxdfUNjLv9A4IgYZMJwytj4xIbCEg0tMazjXNJLKxI-AKDQsKF+EgilDiMbYEwTKdPZ+dKCUlR0dHXU8iotAYTFYbhEwAI2ggCn4Sgwpj+pBoAhQ8MRLBRAHpMVAAALAGwAWmikH0xIIBDEBFRCIWn04WJx+KJJIGwHJlOpaLpSL4sJQdxkD0wACZntMznMed8fti8QTiVFSezCJy0D8-iQAUwgewIAJhFAIVCYSggA)

The gist of this is that we **preserve** the type of the implementation function but we check if it's valid according to our rules. However, I'm not sure when this would be needed/helpful in our case - probably a full example involving those filters and a `styled` call would help me to see this (maybe it's already in those branches of yours, I didn't yet have a chance to look there though)